### ISSUE TITLE
Unable to load shared library 'libFabricCommon.so' or one of its dependencies.

### ISSUE THREAD

**Issue by MedAnd** [Azure/service-fabric-issues#1203]
Attempting to port an existing Service Fabric Windows .Net Core solution to Ubuntu Linux fails to deploy with:

Unhandled Exception: System.TypeInitializationException: The type initializer for 'System.Fabric.Common.AppTrace' threw an exception. ---> System.DllNotFoundException: Unable to load shared library 'libFabricCommon.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibFabricCommon.so.so: cannot open shared object file: No such file or directory
   at System.Fabric.Interop.Utility.RunInMTA[TResult](Func`1 func)
   at System.Fabric.Common.Tracing.TraceConfig.InitializeFromConfigStore(Boolean forceUpdate)
   at System.Fabric.Common.AppTrace..cctor()
   --- End of inner exception stack trace ---
   at System.Fabric.Interop.Utility.RunInMTA[TResult](Func`1 func)
   at System.Fabric.Interop.Utility.WrapNativeAsyncInvokeInMTA[TResult](Func`2 beginFunc, Func`2 endFunc, InteropExceptionTracePolicy tracePolicy, CancellationToken cancellationToken, String functionTag)
   at System.Fabric.FabricRuntime.NativeFabricRuntimeFactory.GetNodeContextAsync(TimeSpan timeout, CancellationToken cancellationToken)
   at Microsoft.ServiceFabric.Services.Runtime.RuntimeContext.GetOrCreateAsync(TimeSpan timeout, CancellationToken cancellationToken)
   at Microsoft.ServiceFabric.Services.Runtime.ServiceRuntime.RegisterServiceAsync(String serviceTypeName, Func`2 serviceFactory, TimeSpan timeout, CancellationToken cancellationToken)

Service Fabric Version: 6.3.119.1
Git repo: [AspNetCore.DataProtection.ServiceFabric
](https://github.com/MedAnd/AspNetCore.DataProtection.ServiceFabric/tree/PersistKeysToServiceFabricServiceUriUpgrade)

**Comment by raunakpandya** [comment]
Can you share the repro steps? How are you building the project and then deploying on Linux cluster?

**Comment by MedAnd** [comment]
1. run build.sh
2. run install.sh

fails on:

`sfctl application provision --application-type-build-path ServiceFabric.DataProtection`

**Comment by ashishnegi** [comment]
@MedAnd Is this dev machine or prod setup where you are deploying ?
In target machine, can you check if libFabricCommon.so file present under `Fabric.Code` folder.

Can you see if adding path to LD_LIBRARY_PATH=$LD_LIBRARY_PATH:<path/to/fabric.code> in [entrypoint.sh](https://github.com/MedAnd/AspNetCore.DataProtection.ServiceFabric/blob/PersistKeysToServiceFabricServiceUriUpgrade/ServiceFabric.DataProtection.Web/PackageRoot/Code/entryPoint.sh#L14) helps ?

```
echo $LD_LIBRARY_PATH >> ld_path.out
LD_DEBUG=all
LD_LIBRARY_PATH=$LD_LIBRARY_PATH:<path/to/fabric.code> dotnet $DIR/ServiceFabric.DataProtection.Web.dll $1 $2
```

Which version of dotnet you are using : `dotnet --info` on target machine ?

Can you share the SF traces at asnegi[at]microsoft.com ? Traces will be present under `logs` folder in `SF cluster folder`.

**Comment by ashishnegi** [comment]
@MedAnd In your `entrypoint.sh` , check `LD_LIBRARY_PATH` variable. It should have `Fabric.Code` folder of the cluster. It is added during cluster deployment.

**Comment by MedAnd** [comment]
Hi @ashishnegi... 

This is a Linux Dev VM with .NET Core SDK (reflecting any global.json):

Version:   2.1.302
Commit:    9048955601
 
Runtime Environment:
OS Name:     ubuntu
OS Version:  16.04
OS Platform: Linux
RID:         ubuntu.16.04-x64
Base Path:   /usr/share/dotnet/sdk/2.1.302/
 
Host (useful for support):
  Version: 2.1.2
  Commit:  811c3ce6c0
 
.NET Core SDKs installed:
  2.1.302 [/usr/share/dotnet/sdk]
 
.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.7 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

**libFabricCommon.so file is present under Fabric.Code folder as per:**

![image](https://user-images.githubusercontent.com/13365135/43774689-59404230-9a8d-11e8-87cb-0b1f803f25c4.png)

Now checking LD_LIBRARY_PATH with the entrypoint.sh file being a copy of generated output from yo azuresfcsharp.

**Comment by MedAnd** [comment]
@ashishnegi - some progress... after I added the below two lines (as per your advice) into the entryPoint.sh file, and also re-structured the location of where I publish the packages to (MUST be directly under ServiceFabric.DataProtection folder) I can use sfctl to upload, provision and create. 

The AspNetCore.DataProtection.ServiceFabric service is running fine however the Kestrel based service fails (ServiceFabric.DataProtection.Web). Logs of the error are on their way to your inbox!

entryPoint.sh file now contains:

> export LD_DEBUG=all 
> export LD_LIBRARY_PATH="/opt/microsoft/servicefabric/bin/Fabric/Fabric.Code"

The LD_LIBRARY_PATH was not globally set with the Service Fabric dynamic library folder path. I also copied the entryPoint.sh file from the yo azuresfcsharp output. Bugs in the Service Fabric install or azuresfcsharp project artifact generation maybe? If the later, should I log a separate issue, or maybe this is documented in the Service Fabric Linux docs?

PS. Last step is to work out why Kestrel service is failing...

**Comment by MedAnd** [comment]
@raunakpandya @ashishnegi - the latest PR addresses most Kestrel issues. Of interest I had to set the same environment Variable (LD_LIBRARY_PATH) for the Debug to work in VS Code.

**Comment by ashishnegi** [comment]
@MedAnd Can you try this on a clean machine ? 
This might be a bug in the installer or someone (program or user) overwrote that variable.

@maburlik Is LD_LIBRARY_PATH set to Fabric.Code as part of Linux installer or hosting sets it for each Linux process before starting it ?

**Comment by ashishnegi** [comment]
@MedAnd I looked at the logs you shared. It loaded the libraries and then seg faulted. You need to look at dump of process to understand where the crash happened.

**Comment by MedAnd** [comment]
Hi @ashishnegi - my last PR addresses most Kestrel issues (self inflicted)... will loop back once I have everything 100%. 

Re: overall experience on Linux, would be good to confirm above issues with LD_LIBRARY_PATH as I do not think this is in the documentation anywhere?

PS. Wondering if yo azuresfcsharp template could also generate the .vscode/launch.json file to make debugging easier...

**Comment by maburlik** [comment]
The LD_LIBRARY_PATH is exported for each root Fabric process. Child processes inherit this settings.

See where this is configured:
https://github.com/Microsoft/service-fabric/blob/master/src/prod/linuxsetup/runtime/scripts/deployfabric.sh
https://github.com/Microsoft/service-fabric/blob/master/src/prod/linuxsetup/runtime/scripts/starthost.sh
https://github.com/Microsoft/service-fabric/blob/master/src/prod/tools/linux/ClusterDeployer/ClusterDeployer.sh
https://github.com/Microsoft/service-fabric/blob/master/src/prod/tools/linux/ClusterDeployer/TraceProcessor.sh
prod\src\Hosting2\ContainerRuntimeHandler.cpp
…

**Comment by ashishnegi** [comment]
@MedAnd Thanks for the yo azuresfcsharp template feedback. If you have launch.json, feel free to open a PR at [generator-azuresfcsharp](https://github.com/Microsoft/generator-azuresfcsharp).

For `LD_LIBRARY_PATH`, it should be set for you by us in your process environmen. see [here](linux : https://github.com/Microsoft/service-fabric/blob/master/src/prod/src/Hosting2/ProcessActivator.Linux.cpp#L474). 

What is its value in your setupentrypoint.sh ?
You can put this in your entry point and see the `ld_path.out` file :

```
echo $LD_LIBRARY_PATH >> ld_path.out
```

**Comment by MedAnd** [comment]
@ashishnegi - have re-checked above and confirm LD_LIBRARY_PATH has the correct value when I check the ld_path.out file created in the /work folder. 

Have updated my entryPoint.sh files accordingly given LD_LIBRARY_PATH is working correctly on Ubuntu when invoked by the Service Fabric bootstrapping process.

**Comment by ashishnegi** [comment]
@MedAnd When you saw the errors that you reported, Were you invoking the service process yourself from command line ? Or creating the app in the SF by deploying it ?

If the `LD_LIBRARY_PATH` is set for you in entrypoint.sh, when you deploy the app, then you should not see this error of unable to find `libFabricCommon.so`.
